### PR TITLE
Implement And Test Post Withdraw Hooks

### LINKED ISSUE #37: Post Withdrawal Hooks
## Problem

When the user withdraws funds from a stealth address, they can provide an address where they want to send them. If the user chooses an address that is tied to their identity, this compromises the privacy benefits of Umbra.

One way to improve on this would be to integrate with other services in some way. A user might want to move funds directly into Tornado cash, or another mixer. They may want to move them directly into a new account on a layer 2 solution, or put them directly into a DeFi protocol of some kind.

Currently, the protocol does not provide an easy way to do this, as the contract simply takes a withdraw address and transfers token ownership to that address.

## Proposal

The protocol could provide an affordance for an optional post-withdrawal hook which would be called after the transfer of the assets. The protocol could also take an optional set of parameters, provided as abi encoded bytes via an argument to the withdraw method, which would be passed to the contract implementing the hook.

Anyone could implement any custom post-withdrawal hook contract with behavior of their choosing. Whatever frontend the user was using to interact with the protocol would specify the withdrawal hook contract, as well as collect relevant data and send it as parameters to the contract

### Option 1: Send Withdraw Hook Contract Address As Parameter

* The withdraw method could have two additional parameters: one for the address of the withdraw-hook contract, and the other for the abi-encoded bytes representing parameters to pass to it.
* The Umbra contract calls the withdraw hook based on one of the following criteria:
  * If the withdraw hook address is not `address(0)`, call the expected method passing the parameters
  * There could simply be a second withdraw method w/o the withdraw hook parameters, and users/frontends choose which one to call base on there needs

### Option 2: Withdraw Hook Contract Is Withdraw Receiver

* The withdraw receiver is assumed to be the contract that has the withdraw hook method to call
* The Umbra contract calls the withdraw hook method on the receiving address based on one of these methods:
  * By checking that the abi-encoded byte parameters are non-zero
  * There could simply be a second withdraw method w/o the withdraw hook parameters, and users/frontends choose which one to call base on there needs

## Considerations

* Reentrancy would be an obvious big risk here, so proper use of checks-effects-interaction and/or reentrancy guards would be critical
* Gas: especially in the case of a user who is not using a post withdrawal hook action, this adds overhead. Would be an argument for a separate method for calls that require the hook.

### PR DESCRIPTION
**Pull request by apbendi**
* Extracts common withdrawal functionality into shared internal methods
* Defines an interface for Umbra post withdraw hooks
* Implements withdraw methods that call this hook after the token withdrawal completes
* Adds tests exercising post withdraw hook usage

TODO:

- [x] Fix `umbra-js` signature methods (and others?)

closes #37

### CODE REVIEW

File: contracts/test/umbra-hook-test.js
Comment (mds1) [inline contracts/test/umbra-hook-test.js:74]: Great data choice 😂

File: contracts/test/umbra-hook-test.js
Comment (mds1) [inline contracts/test/umbra-hook-test.js:95]: Does it matter if this is passed is as an empty array vs empty string? I'd have thought `""` would be the way to do this, so curious if both work to pass in empty bytes. Surprised it's returned as `null` and not an empty string, maybe that's a web3 convention? Doesn't matter much, just curious

File: contracts/test/umbra-hook-test.js
Comment (mds1) [inline contracts/test/umbra-hook-test.js:152]: Found an extra space 😈

Which reminds me—before merging, just run `yarn prettier` from the repo root just in case that wasn't fixed properly

File: contracts/test/umbra-hook-test.js
Comment (mds1) [inline contracts/test/umbra-hook-test.js:55]: Another nit: Do you mind using additional `describe` blocks to group the below tests where it makes sense to? A bit hard to follow the different combinations of hook/no hook, data/no data, meta-tx/regular-tx, etc. I think grouping might make it easier to make sure all the combinations are covered

File: contracts/test/umbra-test.js
Comment (mds1) [inline contracts/test/umbra-test.js:189]: It seems this error string was not changed in the code, so confused how these tests were passing before?

File: contracts/test/umbra-test.js
Comment (mds1) [inline contracts/test/umbra-test.js:394]: Why was this changed from `relayerTokenFee` to `bigFee`?

File: umbra-js/src/classes/Umbra.ts
Comment (mds1) [inline umbra-js/src/classes/Umbra.ts:440]: A few things on this section:

1. Ah, so maybe the `[]` over `""` is not just a web3 thing! Maybe just set the default data value to `[]` so we can skip the first check and just initialize with `let dataArg = data`? Guessing you didn't do that because TS complained about `[]` as a type, though

2. Prefer to use `!isHexString` here to be more robust instead of checking the first two characters are 0x

3. Seems error message should be `Data input must be null or a string in hex form with 0x prefix`

3. Seems we could tidy things up a bit to remove some branches and only have one `if` condition that checks for a valid hex string. If you can't use `[]` as the default value, and are feeling emboldened to use some ternary, you could even just do the default input as `data: string = "0x"` then in the method do:

```typescript
if (typeof data !== 'string' || !isHexString(data)) {
  throw new Error('...')
}
const dataArg = data === '0x' ? [] : data;
```

**Review by mds1 — CHANGES_REQUESTED** [review]
Really great work! I like this. A bunch of comments but no showstoppers.

After running `yarn` at the root, then `cd contracts` and `yarn test`, my tests fail pretty quickly:

```
yarn run v1.22.10
$ concurrently --success first --kill-others "hardhat node &>/dev/null" "hardhat test --network localhost"
[1] Creating Typechain artifacts in directory typechain for target ethers-v5
[0] hardhat node &>/dev/null exited with code 1
--> Sending SIGTERM to other processes..
[1] hardhat test --network localhost exited with code SIGTERM
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

------------------------

Also, I haven't thought about this much yet, but worth discussing is how this approach compares to just adding `target`, and `data` input parameters and having `Umbra.sol` do the call directly, instead of requiring an external call that uses an interface. And I this leads to a similar question of—why do ERCs 677, 777, and 1363 all take this approach instead of a `target` and `data` approach? 

I *think* this current approach is the way to go, and is more flexible, because you do have the hook interface option, and if you don't want to use that, just make the hook recipient a generic multi-call like contract to do anything (this comes at the expense of some extra gas, though, due to the intermediary contract)

Also, this is similar enough to the above ERCs that, if we stick with this approach, there may already be a similar, non-token specific ERC we should use instead of basically writing our own. Or we could even just exactly follow the conventions of one of those above ERCs where it makes sense to, for method names, etc.

**Comment by apbendi** [comment]
> Also, I haven't thought about this much yet, but worth discussing is how this approach compares to just adding target, and data input parameters and having Umbra.sol do the call directly

I think the primary reason to use a pre-defined interface, instead of just arbitrary target + call data, is that without the interface, you can't develop a receiving contract that can act trustlessly on the withdrawal parameters it receives. For example, the receiving contract wouldn't know which token or what amount of it had been withdrawn.

> Also, this is similar enough to the above ERCs that, if we stick with this approach, there may already be a similar, non-token specific ERC we should use instead of basically writing our own. Or we could even just exactly follow the conventions of one of those above ERCs where it makes sense to, for method names, etc.

Well, given that we have different/additional parameters to consider, I think it *has* to be different from other standards, which have their own values to forward. Agree with looking at those standards and following similar conventions where possible! I actually based the implementation off of ERC777 but will give them all another look over.

File: contracts/contracts/UmbraHookable.sol
Comment (apbendi) [reply contracts/contracts/UmbraHookable.sol:6]: I'm very open to naming suggestions here as `UmbraHookable` was my the least-worst option I could come up with. That said, I intentionally avoided `recipient` because it felt overloaded and collided with other concepts in the protocol, i.e. the receiver of the stealth payment.

As for `callHook`, now that I think about it, using the past test description of the action that occurred is probably better. So in this case that would be something like `tokensWithdrawn` I think.

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol:171]: So is your suggestion to basically re-implement/copy paste the withdrawal logic across the two methods? This is how it was previously, but I combined them in this PR because we now had 4 separate methods to withdraw from and it felt appropriate to consolidate the logic. That said, perhaps I'll explore 2 internal implementations as a gas saving technique as part of a gas-golfing pass.

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol:257]: Hmm, yep, I think you're right. Since we no longer explicitly prevent stealth address re-use, they could re-enter by first sending more tokens to the same address. Shucks, I thought this was a useful side effect.

I also think you're right about the impact of re-entrance here, but this is probably something we want to flag explicitly for the auditors.

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol:245]: Not opposed to it, but what would the benefit be? Most/all of the content would just be duplicated from the `external` method calls, no?

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol:271]: I considered this. I left it out because of gas considerations, and because the hook itself could always emit an event if it needed something recorded. Do you think it's worth gas expenses?

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol:264]: > If I want the hook to do something with my tokens, the hook address and acceptor address need to be the same

I don't think this is necessarily true (though it would probably be common). You could, for example, have a hook contract that is already pre-approved by the acceptor for a `transferFrom`, as one contrived example.

File: contracts/test/umbra-hook-test.js
Comment (apbendi) [reply contracts/test/umbra-hook-test.js:74]: Start your morning right 🥩 ☕ !

File: contracts/contracts/Umbra.sol
Comment (mds1) [reply contracts/contracts/Umbra.sol:171]: It was just a thought since this felt slightly dirty, though it does work fine. I don't feel strongly, so up to you which approach you want to use

File: contracts/contracts/UmbraHookable.sol
Comment (mds1) [reply contracts/contracts/UmbraHookable.sol:6]: True, good point on the overloaded use of "recipient". I do like the `tokensWithdrawn` name, but am stuck coming up with other names for `UmbraHookable`. @wildmolasses help!

File: contracts/contracts/Umbra.sol
Comment (mds1) [reply contracts/contracts/Umbra.sol:245]: Yep, would be duplicated basically. Just easier for reading/auditing IMO, so you don't need to jump around to find the docs. If we don't do that, should at least point the reader to where they can find docs for the function inputs

File: contracts/contracts/Umbra.sol
Comment (mds1) [reply contracts/contracts/Umbra.sol:271]: > the hook itself could always emit an event if it needed something recorded

True, though that doesn't help much with general chain analytics to see, e.g. what % of withdraws call a hook

Maybe we add this to the discussion as part of #119?

File: contracts/contracts/Umbra.sol
Comment (mds1) [reply contracts/contracts/Umbra.sol:264]: Hm, yea good point. The main thought I had here was this part though:

> my first impression when reviewing this PR was the the hook would always be the same as the acceptor, if using a hook contract

But since that isn't true, still might be worth documenting somewhere, since the functionality wasn't clear until I reached this line. Guess this is really just a documentation nit 😛

File: contracts/test/umbra-hook-test.js
Comment (apbendi) [reply contracts/test/umbra-hook-test.js:95]: yeah, I wasted time figuring this one out. Passing `""` to ethers' `defaultAbiCoder.encode` function results in the following error AND to web3.js as a contract call argument both result in the following error:

```
Error: invalid arrayify value (argument="value", value="", code=INVALID_ARGUMENT, version=bytes/5.0.10)
```

Given that they're the same, I assume some underlying dependency requires this. I don't like it and wish it would use `""` or just `null` but this is what it's looking for.

File: contracts/test/umbra-hook-test.js
Comment (apbendi) [reply contracts/test/umbra-hook-test.js:152]: I don't know how you catch these!

I'd like to hold off on running prettier to avoid a bunch of unrelated diff noise on this PR. Will open a separate PR for that after this one is merged.

I also don't understand why it's apparently not running on the commit hooks, despite it appearing to do something 🤷‍♂️

File: contracts/test/umbra-hook-test.js
Comment (apbendi) [reply contracts/test/umbra-hook-test.js:55]: Good idea!

File: contracts/test/umbra-test.js
Comment (apbendi) [reply contracts/test/umbra-test.js:189]: When I added a shared internal withdraw message, I could no longer have a more specific error message for the non-meta withdrawal method, so the expected error for this test case changed, even though the behavior didn't.

File: contracts/test/umbra-test.js
Comment (apbendi) [reply contracts/test/umbra-test.js:394]: Good question! Apparently, this test was reverting with the right error but for the wrong reason. It *should* have been using `bigFee` all along. Refactoring the contract internals exposed the incorrect test, so I had to fix it.

File: umbra-js/src/classes/Umbra.ts
Comment (apbendi) [reply umbra-js/src/classes/Umbra.ts:440]: > Ah, so maybe the [] over "" is not just a web3 thing! Maybe just set the default data value to [] so we can skip the first check and just initialize with let dataArg = data? Guessing you didn't do that because TS complained about [] as a type, though

See above— same error for both web3 and ethers.

> Prefer to use !isHexString here to be more robust instead of checking the first two characters are 0x
>
> Seems error message should be Data input must be null or a string in hex form with 0x prefix


Good call, on both. Will update.

> If you can't use [] as the default value

So, I'm not actually sure if I can use `[]` because I didn't try. As mentioned, I really disliked the underlying libraries requiring `[]` for null data, so I wrote our public method intentionally to accept `null` or a hex string.

> are feeling emboldened to use some ternary, you could even just do the default input as data: string = "0x" then in the method do

I find this harder to read and less clear as a public interface 😬 . I think the caller should be able to pass `null` if they want, and if they don't, it should default to no data.

**Comment by apbendi** [comment]
> After running yarn at the root, then cd contracts and yarn test, my tests fail pretty quickly:

Is it possible you have another local network running somewhere? I think that might be the issue, but not sure.

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol:171]: Cool, gonna leave for now, but will do some exploration while gas golfing later this week.

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol]: Yeah, I agree it would be useful to some extent, but it's the same as indexing the params: is the use worth the gas costs. Cool, added to that PR for further investigation.

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol]: Fair enough! Added to both internal methods.

File: contracts/contracts/Umbra.sol
Comment (apbendi) [reply contracts/contracts/Umbra.sol:273]: Hmm, I think I'm actually confused about what you're confused about 🙃 

Can you clarify what you mean here? You're not saying the hook *should* be the same as the acceptor, but you thought it had to be? What gave you that impression, if so, and what would you add to the docs to make it clearer?

**Comment by mds1** [comment]
> Is it possible you have another local network running somewhere? I think that might be the issue, but not sure.

You were right, I did, though that didn't resolve the issue unfortunately. If I run `yarn hardhat node` and `yarn hardhat test --network localhost` in two different terminal windows, everything works fine and tests pass, but now if I run `yarn test` I get the below error. @wildmolasses Are you able reproduce this / have any ideas?

```
yarn run v1.22.10
$ concurrently --success first --kill-others "hardhat node &>/dev/null" "hardhat test --network localhost"
[1] Creating Typechain artifacts in directory typechain for target ethers-v5
[1] Successfully generated Typechain artifacts!
[1] web3-shh package will be deprecated in version 1.3.5 and will no longer be supported.
[1] web3-bzz package will be deprecated in version 1.3.5 and will no longer be supported.
[1] Error HH108: Cannot connect to the network localhost.
[1] Please make sure your node is running, and check your internet connection and networks config
[1] For more info go to https://hardhat.org/HH108 or run Hardhat with --show-stack-traces
[1] 
[1] hardhat test --network localhost exited with code 1
--> Sending SIGTERM to other processes..
[0] hardhat node &>/dev/null exited with code SIGTERM
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: contracts/contracts/Umbra.sol
Comment (mds1) [reply contracts/contracts/Umbra.sol]: Did we ever create an issue to track notes for auditors? We had a few other things we wanted to mention, though I can't recall what they were at the moment 😬

File: contracts/contracts/Umbra.sol
Comment (mds1) [reply contracts/contracts/Umbra.sol:273]: > You're not saying the hook should be the same as the acceptor, but you thought it had to be? 

Correct, that's why I thought when I started reviewing the PR, but I don't know why I thought that! I think it's fine to just move on though since from the interface makes it clear the acceptor is an explicit parameter

File: contracts/test/umbra-hook-test.js
Comment (mds1) [reply contracts/test/umbra-hook-test.js]: Ahhh, I think it's because bytes are usually represented in byte arrays where `0x1234` is represented as two bytes of `[18, 52]`. See:

- https://docs.ethers.io/v5/single-page/#/v5/api/utils/bytes/-%23-utils-arrayify
- https://docs.ethers.io/v5/single-page/#/v5/api/signer/-%23-Signer--signing-methods

File: umbra-js/src/classes/Umbra.ts
Comment (mds1) [reply umbra-js/src/classes/Umbra.ts]: > So, I'm not actually sure if I can use [] because I didn't try. As mentioned, I really disliked the underlying libraries requiring [] for null data, so I wrote our public method intentionally to accept null or a hex string.

Give our new understanding of why they both use `[]`, this might be more acceptable now and we may want to allow it

File: contracts/contracts/UmbraHookable.sol
Comment (mds1) [reply contracts/contracts/UmbraHookable.sol:6]: Let's also add an `I` prefix to the interface name

File: contracts/contracts/Umbra.sol
Comment (mds1) [reply contracts/contracts/Umbra.sol]: Now tracked in #123

File: contracts/test/umbra-hook-test.js
Comment (mds1) [reply contracts/test/umbra-hook-test.js]: It seems passing in `0x` also works

File: umbra-js/src/classes/Umbra.ts
Comment (apbendi) [reply umbra-js/src/classes/Umbra.ts]: Updated with ecec84a6d4a55c2a17e07cf2403260c38947b232

File: contracts/test/umbra-hook-test.js
Comment (apbendi) [reply contracts/test/umbra-hook-test.js]: Updated with ecec84a6d4a55c2a17e07cf2403260c38947b232

File: contracts/contracts/UmbraHookable.sol
Comment (apbendi) [reply contracts/contracts/UmbraHookable.sol]: Renamed to `IUmbraHookReceiver`

**Pull request merged**
